File: Signal.CPP/BiSpcAnalyzer.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "BiSpcAnalyzer.h"
#include "FileSignal.hpp"
#include "VCLTee.TeeStore.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TBiSpcAnalyzerForm *BiSpcAnalyzerForm;
//---------------------------------------------------------------------------
__fastcall TBiSpcAnalyzerForm::TBiSpcAnalyzerForm(TComponent* Owner)
	: TForm(Owner)
{
	SignalRead1->FileName = ExtractFileDir(Application->ExeName) + "\\bz.sfs";
	BiSpectrumAnalyzer->Amplt->SetZero();
	FrequencyEditChange(this);
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::SpectrumEditButtonClick(TObject *Sender)
{
	SpectrumAnalyzerDialog->Execute();	
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::ChartEditButtonClick(TObject *Sender)
{
	ChartEditor->Execute();	
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::SpectrumAnalyzerParameterUpdate(
      TObject *Sender)
{
	SpectrumAnalyzer->Update();
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::Panel2Resize(TObject *Sender)
{
	SpectrumChart1->Height = Panel2->Height / 2;
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::Openfile1Click(TObject *Sender)
{
	OpenDialog->Filter = SignalDialogFilter;
	if (OpenDialog->Execute())
	{
		SignalRead1->FileName = OpenDialog->FileName;
		SignalRead1->OpenFile();
	}
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::ToolButton1Click(TObject *Sender)
{
	BiSpectrumAnalyzerDialog->Execute();	
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::SaveConfigClick(TObject *Sender)
{
	SaveDialog->Filter = "Config files (*.cfg)|*.cfg";
	if (SaveDialog->Execute())
	{
		TFileStream *a = new TFileStream(SaveDialog->FileName, fmOpenWrite | fmCreate);
		try
		{
			//MtxSaveChartToStream(TCustomChart(SpectrumChart),a);
			//MtxSaveChartToStream(TCustomChart(SpectrumChart1),a);
			SaveChartToStream((TCustomChart*)SpectrumChart,a);
			SaveChartToStream((TCustomChart*)SpectrumChart1,a);
			SpectrumAnalyzer->SaveTemplateToStream(a);
			BiSpectrumAnalyzer->SaveTemplateToStream(a);
		}
		__finally
		{
			delete a;
		}
	}
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::LoadConfigClick(TObject *Sender)
{
	OpenDialog->Filter = "Config files (*.cfg)|*.cfg";
	if (OpenDialog->Execute())
	{
		TFileStream *a = new TFileStream(OpenDialog->FileName, fmOpenRead);
		try
		{
			//MtxLoadChartFromStream(TCustomChart(SpectrumChart),a);
			//MtxLoadChartFromStream(TCustomChart(SpectrumChart1),a);

	 //		LoadChartFromStream((TCustomChart*)SpectrumChart,a);
//			LoadChartFromStream((TCustomChart*)SpectrumChart,a);
			SpectrumAnalyzer->LoadTemplateFromStream(a);
			BiSpectrumAnalyzer->LoadTemplateFromStream(a);
			(*SpectrumChart->Spectrums)[0]->Input = SpectrumAnalyzer;
			(*SpectrumChart1->Spectrums)[0]->Input = BiSpectrumAnalyzer;
		}
		__finally
		{
			delete a;
		}
	}
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::BiSpectrumAnalyzerParameterUpdate(
      TObject *Sender)
{
	((TColorLineTool*)(SpectrumChart->GetTool(2)))->Value = BiSpectrumAnalyzer->BiAnalyzer->Frequency;

	BiSpectrumAnalyzer->UpdateSpectrum();
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::Recalculate1Click(TObject *Sender)
{
	Screen->Cursor = crHourGlass;
	try
	{
		SignalRead1->Length = 1024;
		SignalRead1->RecordPosition = 0;
		SignalRead1->LastFrameCheck = lfcLastFullBlock;
		SignalRead1->OverlappingPercent = 0;
		//just make sure that the processing does not take too long
		SignalRead1->SelectionStart = 0;
		SignalRead1->SelectionStop = 50000;
		SpectrumAnalyzer->Update();
		BiSpectrumAnalyzer->ResetAveraging();
		while (BiSpectrumAnalyzer->Pull() != pipeEnd) // do averaging
		{
			ProgressLabel->Caption = "Progress: " + FormatSample("0",SignalRead1->FrameNumber*100/SignalRead1->MaxFrames) + " [%]";
			Update();
		}
		BiSpectrumAnalyzer->BiAnalyzer->Update(); //compute bicoherence
		BiSpectrumAnalyzer->UpdateSpectrum();
	}
	__finally
	{
		Screen->Cursor = crDefault;
	}
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::FrequencyEditChange(TObject *Sender)
{
	BiSpectrumAnalyzer->BiAnalyzer->Frequency = FrequencyEdit->Position;
	BiSpectrumAnalyzerParameterUpdate(Sender);
}
//---------------------------------------------------------------------------
void __fastcall TBiSpcAnalyzerForm::Copy1Click(TObject *Sender)
{
//	((TSpectrumMarkTool*)((*SpectrumChart1->Tools)[1]))->CursorActive = false;
//	SpectrumChart1->CopyToClipboardMetafile(true,Rect(0,0,750,400));
//	((TSpectrumMarkTool*)((*SpectrumChart1->Tools)[1]))->CursorActive = true;

	((TSpectrumMarkTool*)(SpectrumChart1->Tools->Items[1]))->CursorActive = false;
	SpectrumChart1->CopyToClipboardMetafile(true,Rect(0,0,750,400));
	((TSpectrumMarkTool*)(SpectrumChart1->Tools->Items[1]))->CursorActive = true;
}
//---------------------------------------------------------------------------

File: Signal.Delphi/BiSpcAnalyzer.pas
unit BiSpcAnalyzer;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, ComCtrls, ToolWin, ExtCtrls, TeeProcs, TeEngine, Chart,
  SignalToolsTee, TeeEdit, SignalToolsDialogs, SpectrumAna, MtxBaseComp,
  SignalTools, SignalAnalysis, Series, StdCtrls, TeeTools, FileSignal,
  MtxDialogs, MtxVec, MtxComCtrls, Menus, TeeGDIPlus;

type
  TBiSpcAnalyzerForm = class(TForm)
    Panel1: TPanel;
    ToolBar1: TToolBar;
    SpectrumEditButton: TToolButton;
    ChartEditButton: TToolButton;
    SpectrumAnalyzerDialog: TSpectrumAnalyzerDialog;
    ChartEditor: TChartEditor;
    SignalRead1: TSignalRead;
    FrequencyEdit: TMtxFloatEdit;
    FrequencyLabel: TLabel;                               
    Panel2: TPanel;                                        
    SpectrumChart: TSpectrumChart;
    Series1: TFastLineSeries;
    Series2: TPointSeries;
    ChartTool1: TAxisScaleTool;
    Splitter1: TSplitter;
    SpectrumChart1: TSpectrumChart;
    FastLineSeries1: TFastLineSeries;
    PointSeries1: TPointSeries;
    AxisScaleTool1: TAxisScaleTool;
    ChartTool2: TSpectrumMarkTool;
    ProgressLabel: TLabel;
    Button1: TToolButton;
    PopupMenu: TPopupMenu;
    OpenDialog: TOpenDialog;
    SaveDialog: TSaveDialog;
    SaveConfig: TMenuItem;
    LoadConfig: TMenuItem;
    N1: TMenuItem;
    Copy1: TMenuItem;
    N2: TMenuItem;
    Openfile1: TMenuItem;
    ToolButton1: TToolButton;
    Recalculate1: TMenuItem;
    BiSpectrumAnalyzerDialog: TBiSpectrumAnalyzerDialog;
    SpectrumAnalyzer: TSpectrumAnalyzer;
    BiSpectrumAnalyzer: TBiSpectrumAnalyzer;
    ChartTool3: TSpectrumMarkTool;
    ChartTool4: TColorLineTool;
    procedure SpectrumEditButtonClick(Sender: TObject);
    procedure ChartEditButtonClick(Sender: TObject);
    procedure SpectrumAnalyzerParameterUpdate(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure Panel2Resize(Sender: TObject);
    procedure Openfile1Click(Sender: TObject);
    procedure ToolButton1Click(Sender: TObject);
    procedure SaveConfigClick(Sender: TObject);
    procedure LoadConfigClick(Sender: TObject);
    procedure BiSpectrumAnalyzerParameterUpdate(Sender: TObject);
    procedure Recalculate1Click(Sender: TObject);
    procedure FrequencyEditChange(Sender: TObject);
    procedure Copy1Click(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  BiSpcAnalyzerForm: TBiSpcAnalyzerForm;

implementation

{$IFDEF CLR}
{$R *.nfm}
{$ELSE}
{$R *.dfm}
{$ENDIF}

uses MtxVecEdit, MtxVecTee, SignalUtils, Math387, TeeStore
    {$IFDEF CLR}
     ,Borland.Vcl.Types
     {$ENDIF}
     ;

procedure TBiSpcAnalyzerForm.SpectrumEditButtonClick(Sender: TObject);
begin
     SpectrumAnalyzerDialog.Execute;
end;

procedure TBiSpcAnalyzerForm.ChartEditButtonClick(Sender: TObject);
begin
    ChartEditor.Execute;
end;

procedure TBiSpcAnalyzerForm.SpectrumAnalyzerParameterUpdate(Sender: TObject);
begin
     SpectrumAnalyzer.Update;
end;

procedure TBiSpcAnalyzerForm.FormCreate(Sender: TObject);
begin
     SignalRead1.IsDouble := DefaultIsDouble; //select processing precision

     SignalRead1.FileName := ExtractFileDir(Application.ExeName) + '\bz.sfs';
     BispectrumAnalyzer.Amplt.SetZero;
     FrequencyEditChange(Sender);
end;


procedure TBiSpcAnalyzerForm.Panel2Resize(Sender: TObject);
begin
      SpectrumChart1.Height := Panel2.Height div 2;
end;

procedure TBiSpcAnalyzerForm.Openfile1Click(Sender: TObject);
begin
     OpenDialog.Filter := FileSignal.SignalDialogFilter;
     if OpenDialog.Execute then
     begin
          SignalRead1.FileName := OpenDialog.FileName;
          SignalRead1.OpenFile;
     end;
end;

procedure TBiSpcAnalyzerForm.ToolButton1Click(Sender: TObject);
begin
     BispectrumAnalyzerDialog.Execute;
end;

procedure TBiSpcAnalyzerForm.SaveConfigClick(Sender: TObject);
var a: TFileStream;
begin
     SaveDialog.Filter := 'Config files (*.cfg)|*.cfg';
     if SaveDialog.Execute then
     begin
          a := TFileStream.Create(SaveDialog.FileName, fmOpenWrite or fmCreate);
          try
//               MtxSaveChartToStream(TCustomChart(SpectrumChart),a);
//               MtxSaveChartToStream(TCustomChart(SpectrumChart1),a);
               SaveChartToStream(TCustomChart(SpectrumChart),a);
               SaveChartToStream(TCustomChart(SpectrumChart1),a);
               SpectrumAnalyzer.SaveTemplateToStream(a);
               BiSpectrumAnalyzer.SaveTemplateToStream(a);               
          finally
               a.Free;
          end;
     end
end;

procedure TBiSpcAnalyzerForm.LoadConfigClick(Sender: TObject);
var a: TFileStream;
    CustomChart: TCustomChart;
begin
     OpenDialog.Filter := 'Config files (*.cfg)|*.cfg';
     if OpenDialog.Execute then
     begin
          a := TFileStream.Create(OpenDialog.FileName, fmOpenRead);
          try
//               MtxLoadChartFromStream(TCustomChart(SpectrumChart),a);
//               MtxLoadChartFromStream(TCustomChart(SpectrumChart1),a);
               CustomChart := TCustomChart(SpectrumChart);
               LoadChartFromStream(CustomChart,a);
               CustomChart := TCustomChart(SpectrumChart1);
               LoadChartFromStream(CustomChart,a);
               SpectrumAnalyzer.LoadTemplateFromStream(a);
               BiSpectrumAnalyzer.LoadTemplateFromStream(a);
               SpectrumChart.Spectrums[0].Input := SpectrumAnalyzer;
               SpectrumChart1.Spectrums[0].Input := BiSpectrumAnalyzer;
          finally
               a.Free;
          end;
     end;
end;

procedure TBiSpcAnalyzerForm.BiSpectrumAnalyzerParameterUpdate(Sender: TObject);
begin
    TColorLineTool(SpectrumChart.Tools[2]).Value := BispectrumAnalyzer.BiAnalyzer.Frequency;
    BispectrumAnalyzer.UpdateSpectrum;
end;

procedure TBiSpcAnalyzerForm.Recalculate1Click(Sender: TObject);
begin
     Screen.Cursor := crHourGlass;
     try
         SignalRead1.Length := 1024;
         SignalRead1.RecordPosition := 0;
         Signalread1.LastFrameCheck := lfcLastFullBlock;
         SignalRead1.OverlappingPercent := 0;
         //just make sure that the processing does not take too long
         SignalRead1.SelectionStart := 0;
         SignalRead1.SelectionStop := 50000;
         SpectrumAnalyzer.Update;
         BispectrumAnalyzer.ResetAveraging;
         while (BispectrumAnalyzer.Pull <> pipeEnd) do  // do averaging
         begin
              ProgressLabel.Caption := 'Progress: ' + FormatSample('0',SignalRead1.FrameNumber/SignalRead1.MaxFrames*100) + ' [%]';
              Update;
         end;
         BispectrumAnalyzer.BiAnalyzer.Update; //compute bicoherence
         BispectrumAnalyzer.UpdateSpectrum;
     finally
         Screen.Cursor := crDefault;
     end;
end;

procedure TBiSpcAnalyzerForm.FrequencyEditChange(Sender: TObject);
begin
    BispectrumAnalyzer.BiAnalyzer.Frequency := FrequencyEdit.Position;
    BiSpectrumAnalyzerParameterUpdate(Sender);    
end;

procedure TBiSpcAnalyzerForm.Copy1Click(Sender: TObject);
begin
     TSpectrumMarkTool(SpectrumChart1.Tools[1]).CursorActive := False;
     SpectrumChart1.CopyToClipboardMetafile(True,Rect(0,0,750,400));
     TSpectrumMarkTool(SpectrumChart1.Tools[1]).CursorActive := True;
end;

initialization
RegisterClass(TBiSpcAnalyzerForm);

end.
